src/material-experimental/mdc-list/testing/selection-list-harness.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {HarnessPredicate} from '@angular/cdk/testing';
+import {MatListHarnessBase} from './list-harness-base';
+import {
+  ListItemHarnessFilters,
+  ListOptionHarnessFilters,
+  SelectionListHarnessFilters
+} from './list-harness-filters';
+import {getListItemPredicate, MatListItemHarnessBase} from './list-item-harness-base';
+
+/** Harness for interacting with a MDC_based selection-list in tests. */
+export class MatSelectionListHarness extends MatListHarnessBase<
+    typeof MatListOptionHarness, MatListOptionHarness, ListOptionHarnessFilters> {
+  /** The selector for the host element of a `MatSelectionList` instance. */
+  static hostSelector = '.mat-mdc-selection-list';
+
+  /**
+   * Gets a `HarnessPredicate` that can be used to search for a `MatSelectionListHarness` that meets
+   * certain criteria.
+   * @param options Options for filtering which selection list instances are considered a match.
+   * @return a `HarnessPredicate` configured with the given options.
+   */
+  static with(options: SelectionListHarnessFilters = {}):
+      HarnessPredicate<MatSelectionListHarness> {
+    return new HarnessPredicate(MatSelectionListHarness, options);
+  }
+
+  _itemHarness = MatListOptionHarness;
+
+  /** Whether the selection list is disabled. */
+  async isDisabled(): Promise<boolean> {
+    return await (await this.host()).getAttribute('aria-disabled') === 'true';
+  }
+
+  /**
+   * Selects all items matching any of the given filters.
+   * @param filters Filters that specify which items should be selected.
+   */
+  async selectItems(...filters: ListOptionHarnessFilters[]): Promise<void> {
+    const items = await this._getItems(filters);
+    await Promise.all(items.map(item => item.select()));
+  }
+
+  /**
+   * Deselects all items matching any of the given filters.
+   * @param filters Filters that specify which items should be deselected.
+   */
+  async deselectItems(...filters: ListItemHarnessFilters[]): Promise<void> {
+    const items = await this._getItems(filters);
+    await Promise.all(items.map(item => item.deselect()));
+  }
+
+  /** Gets all items matching the given list of filters. */
+  private async _getItems(filters: ListOptionHarnessFilters[]): Promise<MatListOptionHarness[]> {
+    if (!filters.length) {
+      return this.getItems();
+    }
+    return ([] as MatListOptionHarness[])
+        .concat(...await Promise.all(
```

Comment:
Apart from casting the type, I don't see why this needs the `concat`. Isn't it the same as returning the `Promise.all` directly?

Reply:
This has also been taken from the existing list harness. I think the reason for it is to flatten the results from `locatorForAll`. I'm switching it to a reduce in both implementations for more readability and best-practices.